File: android/app/src/main/kotlin/com/goalmaker/app/application/planning/GoalList.kt
package com.goalmaker.app.application.planning

import com.goalmaker.app.application.sync.Replica
import com.goalmaker.app.domain.sync.SyncedTable
import java.time.LocalDate
import java.util.Locale
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull

/**
 * The owner's goals and the amounts logged on them, read from the replica and changed through its
 * outbox (docs/goals.md). The methods block on disk, so callers run them off the main thread.
 */
class GoalList(
    private val replica: Replica,
    private val rows: NewRows,
    private val requestSync: () -> Unit,
) {
    /** Every goal that isn't deleted, longest period first, then by period, position and title. */
    fun all(): List<GoalItem> = replica.all(TABLE).map(::toItem).filterNot(GoalItem::deleted).sortedWith(ORDER)

    fun find(id: String): GoalItem? = replica.get(TABLE, id)?.let(::toItem)?.takeUnless(GoalItem::deleted)

    /** The amounts logged on every goal, not deleted. */
    fun entries(): List<GoalEntryItem> = replica.all(ENTRIES).map(::toEntry).filterNot(GoalEntryItem::deleted)

    /** [all] and [entries], again after every change to either. Collect it off the main thread. */
    fun watch(): Flow<Pair<List<GoalItem>, List<GoalEntryItem>>> =
        combine(replica.watch(TABLE), replica.watch(ENTRIES)) { _, _ -> all() to entries() }

    /** Adds a goal. Null when the draft isn't one the server would keep (see [check]). */
    fun add(draft: GoalDraft): GoalItem? {
        val clean = check(draft) ?: return null
        val siblings = all().count { it.horizon == clean.horizon && it.periodStart == clean.periodStart }
        val row = rows.create(TABLE, values(clean) + ("status" to JsonPrimitive(GoalRules.OPEN)) + ("position" to JsonPrimitive(siblings.toDouble())))
            ?: return null
        replica.queue(TABLE, row)
        requestSync()
        return toItem(row)
    }

    /** Changes a goal to what [draft] says. False when the draft isn't valid or the goal is gone. */
    fun update(id: String, draft: GoalDraft): Boolean {
        val clean = check(draft) ?: return false
        return change(id) { values -> values.putAll(values(clean)) }
    }

    /** Open, done (with the time) or dropped. */
    fun setStatus(id: String, status: String): Boolean {
        if (status !in setOf(GoalRules.OPEN, GoalRules.DONE, GoalRules.DROPPED)) return false
        return change(id) { values ->
            values["status"] = JsonPrimitive(status)
            values["completed_at"] = if (status == GoalRules.DONE) JsonPrimitive(rows.timestamp()) else JsonNull
        }
    }

    /** Deletes a goal softly. Tasks and child goals that pointed at it just have no goal from then on. */
    fun delete(id: String): Boolean = change(id) { values -> values[SyncedTable.DELETED_AT] = JsonPrimitive(rows.timestamp()) }

    /** Logs [amount] on a numeric goal for [day] ("+5 km"); a negative amount corrects. Null for zero. */
    fun logAmount(goalId: String, day: LocalDate, amount: Double): GoalEntryItem? {
        if (amount == 0.0 || amount.isNaN() || amount.isInfinite() || find(goalId) == null) return null
        val row = rows.create(
            ENTRIES,
            mapOf("goal_id" to JsonPrimitive(goalId), "day" to JsonPrimitive(day.toString()), "amount" to JsonPrimitive(amount)),
        ) ?: return null
        replica.queue(ENTRIES, row)
        requestSync()
        return toEntry(row)
    }

    /**
     * Copies the goals of the [horizon] period before the one starting on [start] into it (docs/goals.md)
     * and returns how many were made. Nothing when the period already has goals.
     */
    fun copyPrevious(horizon: GoalHorizon, start: LocalDate): Int {
        val goals = all()
        if (goals.any { it.horizon == horizon && it.periodStart == start }) return 0
        val previous = GoalRules.periodStart(horizon, start.minusDays(1))
        val copies = GoalRules.copies(
            goals.filter { it.horizon == horizon && it.periodStart == previous },
            goals.associateBy(GoalItem::id),
            horizon,
            start,
        )
        replica.inTransaction {
            copies.forEach { copy ->
                add(GoalDraft(copy.title, horizon, start, copy.mode, copy.emoji, copy.parentId, copy.target, copy.unit))
            }
        }
        return copies.size
    }

    // The draft as the server will take it, or null: a title, a numeric goal with a positive target.
    private fun check(draft: GoalDraft): GoalDraft? {
        val title = draft.title.trim().take(MAX_TITLE)
        if (title.isEmpty()) return null
        if (draft.mode !in setOf(GoalRules.MODE_DONE, GoalRules.MODE_TASKS, GoalRules.MODE_NUMBER)) return null
        val number = draft.mode == GoalRules.MODE_NUMBER
        val target = draft.target?.takeIf { number }
        if (number && (target == null || target <= 0.0 || target.isNaN() || target.isInfinite())) return null
        val start = GoalRules.periodStart(draft.horizon, draft.periodStart)
        val parent = draft.parentId?.let(::find)?.takeIf { GoalRules.canServe(draft.horizon, start, it.horizon, it.periodStart) }
        return draft.copy(
            title = title,
            periodStart = start,
            emoji = draft.emoji?.trim()?.takeIf(String::isNotEmpty)?.take(MAX_EMOJI),
            parentId = parent?.id,
            target = target,
            unit = draft.unit?.trim()?.takeIf { number && it.isNotEmpty() }?.take(MAX_UNIT),
        )
    }

    private fun values(draft: GoalDraft): Map<String, JsonElement> = mapOf(
        "title" to JsonPrimitive(draft.title),
        "emoji" to (draft.emoji?.let(::JsonPrimitive) ?: JsonNull),
        "horizon" to JsonPrimitive(draft.horizon.id),
        "period_start" to JsonPrimitive(draft.periodStart.toString()),
        "parent_id" to (draft.parentId?.let(::JsonPrimitive) ?: JsonNull),
        "progress_mode" to JsonPrimitive(draft.mode),
        "target" to (draft.target?.let(::JsonPrimitive) ?: JsonNull),
        "unit" to (draft.unit?.let(::JsonPrimitive) ?: JsonNull),
    )

    private fun change(id: String, edit: (MutableMap<String, JsonElement>) -> Unit): Boolean {
        val row = replica.get(TABLE, id)?.takeIf { it[SyncedTable.DELETED_AT].let { value -> value == null || value == JsonNull } }
            ?: return false
        val values = LinkedHashMap(row)
        edit(values)
        replica.queue(TABLE, JsonObject(values))
        requestSync()
        return true
    }

    private fun toItem(row: JsonObject) = GoalItem(
        id = row.text(SyncedTable.ID).orEmpty(),
        title = row.text("title").orEmpty(),
        horizon = GoalHorizon.of(row.text("horizon")) ?: GoalHorizon.WEEK,
        periodStart = row.text("period_start")?.let(LocalDate::parse) ?: LocalDate.EPOCH,
        mode = row.text("progress_mode") ?: GoalRules.MODE_DONE,
        status = row.text("status") ?: GoalRules.OPEN,
        emoji = row.text("emoji"),
        parentId = row.text("parent_id"),
        target = (row["target"] as? JsonPrimitive)?.doubleOrNull,
        unit = row.text("unit"),
        completedAt = row.text("completed_at"),
        position = (row["position"] as? JsonPrimitive)?.doubleOrNull ?: 0.0,
        deleted = row.text(SyncedTable.DELETED_AT) != null,
    )

    private fun toEntry(row: JsonObject) = GoalEntryItem(
        id = row.text(SyncedTable.ID).orEmpty(),
        goalId = row.text("goal_id").orEmpty(),
        day = row.text("day")?.let(LocalDate::parse) ?: LocalDate.EPOCH,
        amount = (row["amount"] as? JsonPrimitive)?.doubleOrNull ?: 0.0,
        deleted = row.text(SyncedTable.DELETED_AT) != null,
    )

    private companion object {
        const val TABLE = "goals"
        const val ENTRIES = "goal_entries"
        const val MAX_TITLE = 200
        const val MAX_EMOJI = 16
        const val MAX_UNIT = 20
        val ORDER = compareByDescending<GoalItem> { it.horizon.rank }
            .thenBy(GoalItem::periodStart)
            .thenBy(GoalItem::position)
            .thenBy { it.title.lowercase(Locale.ROOT) }

        fun JsonObject.text(name: String): String? = (this[name] as? JsonPrimitive)?.takeUnless { it == JsonNull }?.content
    }
}

File: windows/src/GoalMaker.Core/Planning/GoalList.cs
using System.Globalization;
using System.Text.Json.Nodes;
using GoalMaker.Core.Sync;

namespace GoalMaker.Core.Planning;

/// <summary>
/// The owner's goals and the amounts logged on them, read from the replica and changed through its
/// outbox (docs/goals.md). Every write asks for a sync.
/// </summary>
public sealed class GoalList
{
    private const string Table = "goals";
    private const string EntryTable = "goal_entries";
    private const int MaxTitle = 200;
    private const int MaxEmoji = 16;
    private const int MaxUnit = 20;
    private readonly IReplica replica;
    private readonly NewRows rows;
    private readonly Action requestSync;

    public GoalList(IReplica replica, NewRows rows, Action requestSync)
    {
        this.replica = replica;
        this.rows = rows;
        this.requestSync = requestSync;
        replica.Changed += (_, table) =>
        {
            if (table is Table or EntryTable)
            {
                Changed?.Invoke(this, EventArgs.Empty);
            }
        };
    }

    /// <summary>Raised after every change to the goals or the amounts logged on them.</summary>
    public event EventHandler? Changed;

    /// <summary>Every goal that isn't deleted, longest period first, then by period, position and title.</summary>
    public IReadOnlyList<GoalItem> All() =>
        [.. replica.All(Table).Select(ToItem).Where(goal => !goal.Deleted)
            .OrderByDescending(goal => goal.Horizon)
            .ThenBy(goal => goal.PeriodStart)
            .ThenBy(goal => goal.Position)
            .ThenBy(goal => goal.Title, StringComparer.OrdinalIgnoreCase)];

    public GoalItem? Find(string id) => replica.Get(Table, id) is { } row && ToItem(row) is { Deleted: false } goal ? goal : null;

    /// <summary>The amounts logged on every goal, not deleted.</summary>
    public IReadOnlyList<GoalEntryItem> Entries() => [.. replica.All(EntryTable).Select(ToEntry).Where(entry => !entry.Deleted)];

    /// <summary>Adds a goal. Null when the draft isn't one the server would keep.</summary>
    public GoalItem? Add(GoalDraft draft)
    {
        if (Check(draft) is not { } clean)
        {
            return null;
        }

        var values = Values(clean);
        values["status"] = GoalRules.Open;
        values["position"] = (double)All().Count(goal => goal.Horizon == clean.Horizon && goal.PeriodStart == clean.PeriodStart);
        if (rows.Create(Table, values) is not { } row)
        {
            return null;
        }

        replica.Queue(Table, row);
        requestSync();
        return ToItem(row);
    }

    /// <summary>Changes a goal to what <paramref name="draft"/> says. False when the draft isn't valid or the goal is gone.</summary>
    public bool Update(string id, GoalDraft draft) => Check(draft) is { } clean && Change(id, row =>
    {
        foreach (var (column, value) in Values(clean))
        {
            row[column] = value;
        }
    });

    /// <summary>Open, done (with the time) or dropped.</summary>
    public bool SetStatus(string id, string status) =>
        status is GoalRules.Open or GoalRules.Done or GoalRules.Dropped && Change(id, row =>
        {
            row["status"] = status;
            row["completed_at"] = status == GoalRules.Done ? rows.Timestamp() : null;
        });

    /// <summary>Deletes a goal softly. Tasks and child goals that pointed at it just have no goal from then on.</summary>
    public bool Delete(string id) => Change(id, row => row[SyncedTable.DeletedAt] = rows.Timestamp());

    /// <summary>Logs <paramref name="amount"/> on a numeric goal for <paramref name="day"/> ("+5 km"); a negative amount corrects. Null for zero.</summary>
    public GoalEntryItem? LogAmount(string goalId, DateOnly day, double amount)
    {
        if (amount == 0 || !double.IsFinite(amount) || Find(goalId) is null)
        {
            return null;
        }

        var row = rows.Create(EntryTable, new Dictionary<string, JsonNode?>
        {
            ["goal_id"] = goalId,
            ["day"] = day.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture),
            ["amount"] = amount,
        });
        if (row is null)
        {
            return null;
        }

        replica.Queue(EntryTable, row);
        requestSync();
        return ToEntry(row);
    }

    /// <summary>
    /// Copies the goals of the <paramref name="horizon"/> period before the one starting on
    /// <paramref name="start"/> into it (docs/goals.md) and returns how many were made. Nothing when the
    /// period already has goals.
    /// </summary>
    public int CopyPrevious(GoalHorizon horizon, DateOnly start)
    {
        var goals = All();
        if (goals.Any(goal => goal.Horizon == horizon && goal.PeriodStart == start))
        {
            return 0;
        }

        var previous = GoalRules.PeriodStart(horizon, start.AddDays(-1));
        var copies = GoalRules.Copies(
            goals.Where(goal => goal.Horizon == horizon && goal.PeriodStart == previous),
            goals.ToDictionary(goal => goal.Id, StringComparer.Ordinal),
            horizon,
            start);
        replica.InTransaction(() =>
        {
            foreach (var copy in copies)
            {
                Add(new GoalDraft(copy.Title, horizon, start, copy.Mode, copy.Emoji, copy.ParentId, copy.Target, copy.Unit));
            }
        });
        return copies.Count;
    }

    // The draft as the server will take it, or null: a title, a numeric goal with a positive target.
    private GoalDraft? Check(GoalDraft draft)
    {
        var title = Clip(draft.Title, MaxTitle);
        if (title is null || draft.Mode is not (GoalRules.ModeDone or GoalRules.ModeTasks or GoalRules.ModeNumber))
        {
            return null;
        }

        var number = draft.Mode == GoalRules.ModeNumber;
        var target = number ? draft.Target : null;
        if (number && (target is not { } value || value <= 0 || !double.IsFinite(value)))
        {
            return null;
        }

        var start = GoalRules.PeriodStart(draft.Horizon, draft.PeriodStart);
        var parent = draft.ParentId is { } id && Find(id) is { } found && GoalRules.CanServe(draft.Horizon, start, found.Horizon, found.PeriodStart)
            ? found.Id
            : null;
        return draft with
        {
            Title = title,
            PeriodStart = start,
            Emoji = Clip(draft.Emoji, MaxEmoji),
            ParentId = parent,
            Target = target,
            Unit = number ? Clip(draft.Unit, MaxUnit) : null,
        };
    }

    // Trimmed and cut to length; null when nothing is left.
    private static string? Clip(string? text, int length) => text?.Trim() is { Length: > 0 } trimmed
        ? trimmed.Length > length ? trimmed[..length] : trimmed
        : null;

    private static Dictionary<string, JsonNode?> Values(GoalDraft draft) => new()
    {
        ["title"] = draft.Title,
        ["emoji"] = draft.Emoji,
        ["horizon"] = GoalRules.Id(draft.Horizon),
        ["period_start"] = draft.PeriodStart.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture),
        ["parent_id"] = draft.ParentId,
        ["progress_mode"] = draft.Mode,
        ["target"] = draft.Target,
        ["unit"] = draft.Unit,
    };

    private static double? Number(JsonNode? node) => node is JsonValue value
        ? value.TryGetValue<double>(out var number) ? number : value.TryGetValue<long>(out var whole) ? whole : null
        : null;

    private static DateOnly Day(JsonNode? node) =>
        (string?)node is { } text ? DateOnly.ParseExact(text, "yyyy-MM-dd", CultureInfo.InvariantCulture) : DateOnly.MinValue;

    private static GoalItem ToItem(JsonObject row) => new(
        (string?)row[SyncedTable.Id] ?? string.Empty,
        (string?)row["title"] ?? string.Empty,
        GoalRules.HorizonOf((string?)row["horizon"]) ?? GoalHorizon.Week,
        Day(row["period_start"]))
    {
        Mode = (string?)row["progress_mode"] ?? GoalRules.ModeDone,
        Status = (string?)row["status"] ?? GoalRules.Open,
        Emoji = (string?)row["emoji"],
        ParentId = (string?)row["parent_id"],
        Target = Number(row["target"]),
        Unit = (string?)row["unit"],
        CompletedAt = (string?)row["completed_at"],
        Position = Number(row["position"]) ?? 0,
        Deleted = row[SyncedTable.DeletedAt] is not null,
    };

    private static GoalEntryItem ToEntry(JsonObject row) => new(
        (string?)row[SyncedTable.Id] ?? string.Empty,
        (string?)row["goal_id"] ?? string.Empty,
        Day(row["day"]),
        Number(row["amount"]) ?? 0,
        row[SyncedTable.DeletedAt] is not null);

    private bool Change(string id, Action<JsonObject> edit)
    {
        if (replica.Get(Table, id) is not { } row || row[SyncedTable.DeletedAt] is not null)
        {
            return false;
        }

        edit(row);
        replica.Queue(Table, row);
        requestSync();
        return true;
    }
}
